### PR TITLE
Audio_Heap OK and Documented

### PR DESCRIPTION
**Pull request by engineer124**
Before opening this PR, ensure the following:
- `./format.sh` was run to apply standard formatting.
- `make` successfully builds a matching ROM.
- No new compiler warnings were introduced during the build process.
    - Can be verified locally by running `tools/warnings_count/check_new_warnings.sh`
- New variables & functions should follow standard naming conventions.
- Comments and variables have correct spelling.
---

This audio file manages the audio heap, a heap that is separate to the rest of the game, and used primarily for the audio sequences, sound fonts, and raw audio samples. Much of this documentation is reliant on OoT with a lot of support from Zelda2774, so many thanks.

There are new new things added to this file in MM compared to OoT. One is the addition of a band-pass filter and a band-stop filter. The details of these can be found in the discord, you can @ me if you need help finding it. The second is a bunch of additional code for initializing and updating audio reverbs. Audio reverbs are still poorly understood even in OoT, so I don't have too many comments about the new code here until reverbs are better understood.

This file has a lot of overlap with Audio_Load https://github.com/zeldaret/mm/pull/620, so it would be best to reference that PR if need-be

### CODE REVIEW

File: include/variables.h [hunk 1]
```diff
@@ -3082,6 +3082,7 @@ extern f64 D_801E0EB0;
 // extern UNK_TYPE4 D_801E1068;
 extern UNK_PTR D_801E10B0;
 extern const s16 gAudioTatumInit[];
+#define gTatumsPerBeat (gAudioTatumInit[1])
```
Comment (AngheloAlf) [inline include/variables.h:3085]: What?
Could you explain what's the purpose of this define?

File: src/code/audio/audio_heap.c [hunk 2]
```diff
@@ -1,111 +1,1576 @@
...
-#pragma GLOBAL_ASM("asm/non_matchings/code/audio_heap/func_8018B8FC.s")
+    for (i = 0; i < gAudioContext.audioBufferParameters.numSequencePlayers; i++) {
+        if (gAudioContext.seqPlayers[i].enabled && gAudioContext.seqPlayers[i].seqId == seqId) {
+            AudioSeq_SequencePlayerDisable(&gAudioContext.seqPlayers[i]);
+        }
+    }
+}
 
-#pragma GLOBAL_ASM("asm/non_matchings/code/audio_heap/func_8018B95C.s")
+void* AudioHeap_WritebackDCache(void* mem, u32 size) {
```
Comment (AngheloAlf) [inline src/code/audio/audio_heap.c:113]: Would make sense using `size_t` for the sizes in this file instead of `u32`s?

File: src/code/audio/audio_heap.c [hunk 3]
```diff
@@ -1,111 +1,1576 @@
...
+    Audio_WritebackDCache(mem, size);
+    if (mem) {}
+    return (void*)((u32)mem + 0x20000000);
```
Comment (AngheloAlf) [inline src/code/audio/audio_heap.c:116]: Do you know the meaning of `0x20000000`?
Also, this looks like a job for `uintptr_t`
```suggestion
    return (void*)((uintptr_t)mem + 0x20000000);
```

File: src/code/audio/audio_heap.c [hunk 4]
```diff
@@ -1,111 +1,1576 @@
...
+    }
+    if (tableType == FONT_TABLE) {
+        AudioHeap_DiscardFont(persistent->entries[persistent->numEntries - 1].id);
+    }
+    table[persistent->entries[persistent->numEntries - 1].id] = 0;
+    persistent->numEntries--;
+}
 
-#pragma GLOBAL_ASM("asm/non_matchings/code/audio_heap/func_8018DD98.s")
+void AudioHeap_InitMainPools(s32 initPoolSize) {
```
Comment (AngheloAlf) [inline src/code/audio/audio_heap.c:271]: Could this be a `size_t` too?

File: src/code/audio/audio_heap.c [hunk 5]
```diff
@@ -1,111 +1,1576 @@
...
+    gAudioContext.permanentCache[index].ptr = ret;
+    if (ret == NULL) {
+        return NULL;
+    }
+    gAudioContext.permanentCache[index].tableType = tableType;
+    gAudioContext.permanentCache[index].id = id;
+    gAudioContext.permanentCache[index].size = size;
+    //! @bug UB: missing return. "ret" is in v0 at this point, but doing an
+    // explicit return uses an additional register.
+    // return ret;
```
Comment (AngheloAlf) [inline src/code/audio/audio_heap.c:1008]: This game is so ducktaped together. I'm surprised it even works

File: src/code/audio/audio_heap.c [hunk 6]
```diff
@@ -1,111 +1,1576 @@
...
+    entry = &sampleBankTable->entries[sampleBankId];
+    change.size = entry->size;
+    change.newMedium = entry->medium;
+    change.newAddr = entry->romAddr;
+
+    fakematch = &change.oldAddr;
+    if ((apply != false) && (apply == true)) {
+        u32 temp = change.newAddr;
+
+        change.newAddr = *fakematch; // = change.oldAddr
```
Comment (AngheloAlf) [inline src/code/audio/audio_heap.c:1343]: Good var name :ok_hand:

File: src/code/audio/audio_heap.c [hunk 7]
```diff
@@ -1,111 +1,1576 @@
...
+        return;
+    }
+
+    entry = &sampleBankTable->entries[sampleBankId];
+    change.size = entry->size;
+    change.newMedium = entry->medium;
+    change.newAddr = entry->romAddr;
+
+    fakematch = &change.oldAddr;
+    if ((apply != false) && (apply == true)) {
```
Comment (AngheloAlf) [inline src/code/audio/audio_heap.c:1340]: Excuse me, what? 
Need to be sure I guess?

**Review by AngheloAlf — COMMENTED** [review]
Why is Audio so complex? Even its heap is hard to follow...

Anyways, good work! just a few comments

File: src/code/audio/audio_heap.c (hunk 5 above)
Comment (louist103) [reply src/code/audio/audio_heap.c:1008]: OOT is going to wrap that return in an `#ifdef AVOID_UB` for GCC. Maybe MM would want to do the same

Comment (engineer124) [reply src/code/audio/audio_heap.c:1008]: I think having it in is fine, but I think I'd prefer to wait until it's in the OoT repo incase there's any changes to the actual implementation of this in the review process. Once that's in though, I can apply it here

File: include/variables.h (hunk 1 above)
Comment (engineer124) [reply include/variables.h:3085]: This was a remnant of OoT and was mainly used before `gAudioTatumInit` was named. `gAudioTatumInit[0]` still represents something unknown so I'm still not sure if `gAudioTatumInit` is a good name or not. Nonetheless, this really shouldn't be in variables.h, so I've moved it into audio_heap.c

File: src/code/audio/audio_heap.c (hunk 6 above)
Comment (engineer124) [reply src/code/audio/audio_heap.c]: Yeah... these temps were better named

**Comment by engineer124** [comment]
In addition to the PR suggestions, there were many more `size_t`'s and `uintptr_t` not commented on that were applied throughout this file. I also added some small docs to the audio buffer

**Review by hensldm — COMMENTED** [review]
Just one comment about the bug tag in `AudioHeap_AllocSampleCache`

**Review by AngheloAlf — COMMENTED** [review]
Some minor stuff

File: src/code/audio/audio_heap.c [hunk 8]
```diff
@@ -846,23 +921,26 @@ s32 AudioHeap_ResetStep(void) {
...
-    gAudioContext.audioBufferParameters.aiFrequency = osAiSetFrequency(gAudioContext.audioBufferParameters.frequency);
+
+    // audio buffer parameters
+    gAudioContext.audioBufferParameters.samplingFreq = spec->samplingFreq;
+    gAudioContext.audioBufferParameters.aiSamplingFreq =
+        osAiSetFrequency(gAudioContext.audioBufferParameters.samplingFreq);
+
     gAudioContext.audioBufferParameters.samplesPerFrameTarget =
-        ((gAudioContext.audioBufferParameters.frequency / gAudioContext.refreshRate) + 0xF) & 0xFFF0;
+        ((gAudioContext.audioBufferParameters.samplingFreq / gAudioContext.refreshRate) + 0xF) & 0xFFF0;
```
[the remark this answers is not in the document]
Comment (engineer124) [reply src/code/audio/audio_heap.c:943]: It does more than remind me of `ALIGN16`, it's exactly the `ALIGN16` operation and the macro works just fine here. Nice catch.

Seems fine to me to use the macro, especially since that's what the code is doing. All it's saying is to make sure samples are processed in batches of 16 bytes at once, which is fairly standard

**Review by AngheloAlf — APPROVED** [review]

File: include/z64audio.h [hunk 9]
```diff
@@ -145,7 +162,7 @@ typedef struct {
     /* 0x00 */ u32 medium : 2;
     /* 0x00 */ u32 unk_bit26 : 1;
     /* 0x00 */ u32 unk_bit25 : 1;
-    /* 0x01 */ u32 size : 24;
+    /* 0x01 */ size_t size : 24;
```
Comment (Kenix3) [inline include/z64audio.h:165]: See Audio Load system PR for discussion on this.

File: include/z64audio.h [hunk 10]
```diff
@@ -784,7 +821,7 @@ typedef struct {
     /* 0x0C */ u8* curRamAddr;
     /* 0x10 */ u8* ramAddr;
     /* 0x14 */ s32 status;
-    /* 0x18 */ s32 bytesRemaining;
+    /* 0x18 */ u32 bytesRemaining;
```
Comment (Kenix3) [inline include/z64audio.h:824]:
```suggestion
    /* 0x18 */ size_t bytesRemaining;
```
size_t is to express a count in size of bytes.

**Review by Kenix3 — CHANGES_REQUESTED** [review]
Great work on this as well! I suspect we'll be documenting this code for years given how complex it is

**Review by Kenix3 — APPROVED** [review]

File: src/code/audio/audio_heap.c [hunk 11]
```diff
@@ -1,111 +1,1688 @@
...
-#pragma GLOBAL_ASM("asm/non_matchings/code/audio_heap/func_8018BA64.s")
+/**
+ * Perform a writeback from the L1 data cache to the ram.
+ */
+void* AudioHeap_WritebackDCache(void* addr, size_t size) {
+    Audio_WritebackDCache(addr, size);
+    if (addr) {}
 
-#pragma GLOBAL_ASM("asm/non_matchings/code/audio_heap/func_8018BB28.s")
+    // KREG0 to KREG1 (ensures data is written straight to ram instead of the data cache)
```
Comment (Thar0) [inline src/code/audio/audio_heap.c:121]:
```suggestion
    // KSEG0 to KSEG1 (ensures data is written straight to ram instead of the data cache)
```

**Review by Thar0 — CHANGES_REQUESTED** [review]
